### PR TITLE
replace trackersourcelinker with trackermeasurement

### PR DESCRIPTION
**Pull request by ShujieL**
### Briefly, what does this PR introduce?
use the new Measurement2D structure to pass tracker hits to CKF, instead of using sourcelinker and meaurements from ACTS. Ideally this will allow measurement2D (eventually trackerhits) to trajectory association. 

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
TBD
### Does this PR change default behavior?
TBD

### CODE REVIEW

File: src/services/io/podio/JEventProcessorPODIO.cc [hunk 1]
```diff
@@ -152,6 +153,13 @@ JEventProcessorPODIO::JEventProcessorPODIO() {
             "HcalEndcapNMergedHits",
             "HcalEndcapNClusters",
             "HcalEndcapNClusterAssociations",
+            "HcalEndcapPRawHits",   // this causes premature exit of eicrecon
```
Comment (wdconinc) [inline src/services/io/podio/JEventProcessorPODIO.cc:156]: Don't add unrelated changes, in particular make sure not to readd lines that were recently removed.

File: src/algorithms/tracking/CKFTracking.cc [hunk 2]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
         std::unique_ptr<edm4eic::TrackParametersCollection>,
         std::vector<ActsExamples::Trajectories*>
     >
-    CKFTracking::process(const ActsExamples::IndexSourceLinkContainer &src_links,
-                         const ActsExamples::MeasurementContainer &measurements,
+    CKFTracking::process(const edm4eic::Measurement2DCollection& meas2Ds,
                          const edm4eic::TrackParametersCollection &init_trk_params) {
 
+        // auto src_links    = std::vector<std::shared_ptr<ActsExamples::IndexSourceLink>>();
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:85]: Remove commented lines.

File: src/algorithms/tracking/CKFTracking.cc [hunk 3]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
...
+        std::list<ActsExamples::IndexSourceLink> sourceLinkStorage;
+        ActsExamples::IndexSourceLinkContainer src_links;
+        src_links.reserve(meas2Ds.size());
+        std::size_t  hit_index = 0;
+
+
+        for (const auto& meas2D : meas2Ds) {
+            // --Create source links - method of trackersourcelinkerrResult
+            // auto sourceLink = std::make_shared<ActsExamples::IndexSourceLink>(meas2D.getSurface(), hit_index);
+            // src_links.emplace_hint(src_links.end(), *sourceLink);
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:100]: Remove commented lines.

File: src/algorithms/tracking/CKFTracking.cc [hunk 4]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
...
+            // since the input is also geometry-ordered, new items can
+            // be added at the end.
+            src_links.insert(src_links.end(), sourceLink);
+            // ---
+            // Create ACTS measurements
+            Acts::Vector2 loc = Acts::Vector2::Zero();
+            auto pos_x = meas2D.getLoc().a;
+            auto pos_y = meas2D.getLoc().b;
+            loc[Acts::eBoundLoc0] = pos_x;
+            loc[Acts::eBoundLoc1] = pos_y;
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:116]: Nothing x,y about these positions, just put `loc[Acts::eBoundLoc0] = meas2D.getLoc().a;`

File: src/algorithms/tracking/CKFTracking.cc [hunk 5]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
...
+
+
+            Acts::SymMatrix2 cov = Acts::SymMatrix2::Zero();
+            cov(0, 0) = meas2D.getCovariance().xx;
+            cov(1, 1) = meas2D.getCovariance().yy;
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:121]: We should also add cross-covariances.

File: src/algorithms/tracking/CKFTracking.cc [hunk 6]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
...
+
+            auto measurement = Acts::makeMeasurement(sourceLink, loc, cov, Acts::eBoundLoc0, Acts::eBoundLoc1);
+            measurements->emplace_back(std::move(measurement));
+
+            hit_index++;
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:126]: Unused?

File: src/algorithms/tracking/CKFTracking.cc [hunk 7]
```diff
@@ -74,10 +79,53 @@ namespace eicrecon {
...
+        // auto src_links    = std::vector<std::shared_ptr<ActsExamples::IndexSourceLink>>();
+
+        // create sourcelink and measurement containers
+        auto measurements = std::make_shared<ActsExamples::MeasurementContainer>();
+
+          // need list here for stable addresses
+        std::list<ActsExamples::IndexSourceLink> sourceLinkStorage;
+        ActsExamples::IndexSourceLinkContainer src_links;
+        src_links.reserve(meas2Ds.size());
+        std::size_t  hit_index = 0;
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:94]: Unused?

File: src/algorithms/tracking/CKFTracking.cc [hunk 8]
```diff
@@ -172,6 +220,9 @@ namespace eicrecon {
                 // The trajectory entry indices and the multiTrajectory
                 const auto& mj        = multiTrajectory->multiTrajectory();
                 const auto& trackTips = multiTrajectory->tips();
+                // const auto& states    = multiTrajectory->states(); // can't find corresponding function anywhere
+
+
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:225]: Remove commented lines.

File: src/algorithms/tracking/CKFTracking.cc [hunk 9]
```diff
@@ -240,9 +293,46 @@ namespace eicrecon {
...
+                            }
+                        }
+                    }
+
+                });
                 acts_trajectories.push_back(std::move(multiTrajectory));
+            }
 
-            } else {
+         else {
```
Comment (wdconinc) [inline src/algorithms/tracking/CKFTracking.cc:335]: Keep on one line.

File: src/algorithms/tracking/TrackerMeasurement.cc [hunk 10]
```diff
@@ -0,0 +1,143 @@
+// Original license from Gaudi algorithm:
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2022 Whitney Armstrong, Sylvester Joosten, Wouter Deconinck, Dmitry Romanov, Shujie Li
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurement.cc:3]: Update names with just yours. This is new work.

File: src/algorithms/tracking/TrackerMeasurement.cc [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+// TODO refactor header when license is clear
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurement.cc:4]: Remove. License is clear and is LGPL-3.0-or-later.

File: src/algorithms/tracking/TrackerMeasurement.cc [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        auto meas2Ds = std::make_unique<edm4eic::Measurement2DCollection>();
+
+        // To do: add clustering to allow forming one measurement from several hits.
+        // For now, one hit = one measurement.
+        for (const auto *hit: trk_hits) {
+
+            Acts::SymMatrix2 cov = Acts::SymMatrix2::Zero();
+            cov(0, 0) = hit->getPositionError().xx * mm_acts * mm_acts; // note mm = 1 (Acts)
+            cov(1, 1) = hit->getPositionError().yy * mm_acts * mm_acts;
+
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurement.cc:64]: Needs cross-covariances.

File: src/algorithms/tracking/TrackerMeasurement.cc [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+            m_log->trace("   surfaceMap size: {}", surfaceMap.size());
+
+            const auto is = surfaceMap.find(vol_id);
+            if (is == m_acts_context->surfaceMap().end()) {
+                m_log->warn(" WARNING: vol_id ({})  not found in m_surfaces.", vol_id );
+                continue;
+            }
+            const Acts::Surface* surface = is->second;
+            // variable surf_center not used anywhere;
+            // auto surf_center = surface->center(Acts::GeometryContext());
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurement.cc:84]: Remove commented lines.

File: src/global/tracking/TrackerMeasurement_factory.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
+// Created by Dmitry Romanov
```
Comment (wdconinc) [inline src/global/tracking/TrackerMeasurement_factory.h:1]: Update name.

File: src/global/tracking/tracking.cc [hunk 15]
```diff
@@ -48,14 +48,17 @@ void InitPlugin(JApplication *app) {
         app));
 
     // Source linker
-    app->Add(new JChainFactoryGeneratorT<TrackerSourceLinker_factory>(
-            {"CentralTrackingRecHits"}, "CentralTrackerSourceLinker"));
+    // app->Add(new JChainFactoryGeneratorT<TrackerSourceLinker_factory>(
+    //         {"CentralTrackingRecHits"}, "CentralTrackerSourceLinker"));
```
Comment (wdconinc) [inline src/global/tracking/tracking.cc:52]: Remove commented lines.

File: src/global/tracking/tracking.cc [hunk 16]
```diff
@@ -48,14 +48,17 @@ void InitPlugin(JApplication *app) {
...
+    app->Add(new JChainFactoryGeneratorT<TrackerMeasurement_factory>(
+            {"CentralTrackingRecHits"}, "CentralTrackerMeasurements"));
 
     app->Add(new JChainMultifactoryGeneratorT<CKFTracking_factory>(
         "CentralCKFTrajectories",
         {
             "InitTrackParams",
-            "CentralTrackerSourceLinker"
+            "CentralTrackerMeasurements"
+            // "CentralTrackerSourceLinker"
```
Comment (wdconinc) [inline src/global/tracking/tracking.cc:61]: Remove commented lines.

File: src/global/tracking/tracking.cc [hunk 17]
```diff
@@ -72,7 +75,8 @@ void InitPlugin(JApplication *app) {
         "CentralCKFSeededTrajectories",
         {
             "CentralTrackSeedingResults",
-            "CentralTrackerSourceLinker"
+            "CentralTrackerMeasurements"
+            // "CentralTrackerSourceLinker"
```
Comment (wdconinc) [inline src/global/tracking/tracking.cc:79]: Remove commented lines.

File: src/algorithms/tracking/CKFTracking.cc (hunk 5 above)
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:121]: The off diagonal elements are all zero. We can probably add them in the future when we know what number to use

File: src/algorithms/tracking/CKFTracking.cc (hunk 6 above)
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:126]: If you mean hit index, it's used to create sourcelink.

File: src/algorithms/tracking/CKFTracking.cc (hunk 7 above)
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:94]: same as above

File: src/algorithms/tracking/TrackerMeasurement.cc (hunk 12 above)
Comment (ShujieL) [reply src/algorithms/tracking/TrackerMeasurement.cc:64]: same as above, no knowledge on off diagonal elements right now

File: src/algorithms/tracking/CKFTracking.cc (hunk 5 above)
Comment (wdconinc) [reply src/algorithms/tracking/CKFTracking.cc:121]: You should add them now here, even if they are zero.

File: src/algorithms/tracking/CKFTracking.cc [hunk 18]
```diff
@@ -20,6 +22,8 @@
 #include <Acts/Utilities/Logger.hpp>
 #include <edm4eic/Cov2f.h>
 #include <edm4eic/Cov3f.h>
+#include <edm4eic/Measurement2D.h>
+#include <edm4eic/Measurement2DCollection.h>
```
[the remark this answers is not in the document]
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:26]: Can you tell me what exactly to be updated?

[the remark this answers is not in the document]
Comment (wdconinc) [reply src/algorithms/tracking/CKFTracking.cc:26]: `.github/iwyu.imp`

[the remark this answers is not in the document]
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:26]: It's still not clear to me what in iwyu.imp needs to be modified... I think we can address this in a separate issue. Can you review this PR and give it a pass?

[the remark this answers is not in the document]
Comment (wdconinc) [reply src/algorithms/tracking/CKFTracking.cc:26]: It's relevant to this issue because this is where the Measurement2D class is introduced.

[the remark this answers is not in the document]
Comment (wdconinc) [reply src/algorithms/tracking/CKFTracking.cc:26]: Did you check out the link at the top of the file?

[the remark this answers is not in the document]
Comment (ShujieL) [reply src/algorithms/tracking/CKFTracking.cc:26]: New lines added. Thanks for the instruction!

**Review by wdconinc — CHANGES_REQUESTED** [review]
`TrackerSourceLinker.h` should also be deleted, I think.

**Comment by ShujieL** [comment]
> `TrackerSourceLinker.h` should also be deleted, I think.

There's no file name "TrackerSourceLinker.h" in this branch...

File: src/algorithms/tracking/TrackerMeasurement.h [hunk 19]
```diff
@@ -2,31 +2,32 @@
...
-    class TrackerSourceLinker {
+    class TrackerMeasurement {
     public:
         void init(const dd4hep::Detector* detector,
                   const dd4hep::rec::CellIDPositionConverter* converter,
                   std::shared_ptr<const ActsGeometryProvider> acts_context,
                   std::shared_ptr<spdlog::logger> logger);
 
-        eicrecon::TrackerSourceLinkerResult *produce(std::vector<const edm4eic::TrackerHit *> trk_hits);
+        // eicrecon::TrackerMeasurment *produce(std::vector<const edm4eic::TrackerHit *> trk_hits);
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurement.h:29]: Remove commented line.

File: src/global/tracking/CKFTracking_factory.cc [hunk 20]
```diff
@@ -53,39 +46,38 @@ void eicrecon::CKFTracking_factory::Init() {
...
-        source_links.emplace_hint(source_links.end(), *sourceLink);
-    }
+    // // Convert vector of source links to a sorted in geometry order container used in tracking
+    // ActsExamples::IndexSourceLinkContainer source_links;
+    // auto measurements_ptr = source_linker_result->measurements;
+    // for(const auto &sourceLink: source_linker_result->sourceLinks){
+    //     // add to output containers. since the input is already geometry-order,
+    //     // new elements in geometry containers can just be appended at the end.
+    //     source_links.emplace_hint(source_links.end(), *sourceLink);
+    // }
```
Comment (wdconinc) [inline src/global/tracking/CKFTracking_factory.cc:64]: Remove commented lines.

File: src/global/tracking/CKFTracking_factory.cc [hunk 21]
```diff
@@ -53,39 +46,38 @@ void eicrecon::CKFTracking_factory::Init() {
...
-    m_log->debug("Diving into tracking...");
+    // if(m_log->level() <= spdlog::level::trace) {
+    //     m_log->trace("Checking Source links: ");
+    //     for(auto sourceLink: source_links) {
+    //         m_log->trace("   index: {:<5} geometryId: {}", sourceLink.get().index(), sourceLink.get().geometryId().value());
+    //     }
+    // // }
+    // m_log->debug("Source links count: {}", source_links.size());
+    // m_log->debug("Measurements count: {}", source_linker_result->measurements->size());
+    // m_log->debug("Diving into tracking...");
```
Comment (wdconinc) [inline src/global/tracking/CKFTracking_factory.cc:75]: Remove commented lines.

File: src/services/io/podio/JEventProcessorPODIO.cc [hunk 22]
```diff
@@ -52,6 +52,7 @@ JEventProcessorPODIO::JEventProcessorPODIO() {
             // All tracking hits combined
             "CentralTrackingRecHits",
             "CentralTrackSeedingResults",
+            "CentralTrackerMeasurement",
```
Comment (wdconinc) [inline src/services/io/podio/JEventProcessorPODIO.cc:55]:
```suggestion
            "CentralTrackerMeasurements",
```

**Review by wdconinc — COMMENTED** [review]
Two remaining comments:
- Instead of `meas2d`, I'd use a longer name for variables, just `measurements` for a collection and `measurement` for one entry. Shorter variable names are less descriptive.
- The algorithm name TrackerMeasurement doesn't say what it *does*. There's a verb missing, like maybe `TrackerHitToMeasurementConverter` would be a more descriptive name.

**Comment by ShujieL** [comment]
> Two remaining comments:
> 
> * Instead of `meas2d`, I'd use a longer name for variables, just `measurements` for a collection and `measurement` for one entry. Shorter variable names are less descriptive.
> * The algorithm name TrackerMeasurement doesn't say what it _does_. There's a verb missing, like maybe `TrackerHitToMeasurementConverter` would be a more descriptive name.

TrackerMeasurement are renamed into TrackerMeasurementFromHits.
The name "measurement(s)" is used for ACTS measurement. So I chose meas2d (short for measurement2d) to avoid confusion.

File: src/algorithms/tracking/TrackerMeasurementFromHits.h [hunk 23]
```diff
@@ -2,31 +2,32 @@
...
-    class TrackerSourceLinker {
+    class TrackerMeasurementFromHits {
     public:
         void init(const dd4hep::Detector* detector,
                   const dd4hep::rec::CellIDPositionConverter* converter,
                   std::shared_ptr<const ActsGeometryProvider> acts_context,
                   std::shared_ptr<spdlog::logger> logger);
 
-        eicrecon::TrackerSourceLinkerResult *produce(std::vector<const edm4eic::TrackerHit *> trk_hits);
+        // eicrecon::TrackerMeasurment *produce(std::vector<const edm4eic::TrackerHit *> trk_hits);
```
Comment (wdconinc) [inline src/algorithms/tracking/TrackerMeasurementFromHits.h:29]:
```suggestion
```

File: src/global/tracking/CKFTracking_factory.cc [hunk 24]
```diff
@@ -36,8 +29,8 @@ void eicrecon::CKFTracking_factory::Init() {
     InitLogger(app, param_prefix, "info");
 
     // Get ACTS context from ACTSGeo service
-    auto acts_service = app->GetService<ACTSGeo_service>();
-    auto dd4hp_service = app->GetService<DD4hep_service>();
+    auto acts_service   = app->GetService<ACTSGeo_service>();
+    auto dd4hep_service = app->GetService<DD4hep_service>();
```
Comment (wdconinc) [inline src/global/tracking/CKFTracking_factory.cc:33]: This is unused, right?
```suggestion
    auto dd4hep_service = app->GetService<DD4hep_service>();
```

File: src/global/tracking/CKFTracking_factory.cc [hunk 25]
```diff
@@ -4,23 +4,16 @@
...
-#include <functional>
-#include <map>
 
-#include "ActsExamples/EventData/IndexSourceLink.hpp"
 #include "CKFTracking.h"
 #include "CKFTrackingConfig.h"
-#include "algorithms/tracking/TrackerSourceLinkerResult.h"
 #include "datamodel_glue.h"
 #include "services/geometry/acts/ACTSGeo_service.h"
 #include "services/geometry/dd4hep/DD4hep_service.h"
```
Comment (wdconinc) [inline src/global/tracking/CKFTracking_factory.cc:19]: If DD4hep_service is unused, then remove the header.

Comment (wdconinc) [reply src/global/tracking/CKFTracking_factory.cc:19]:
```suggestion
```

**Comment by wdconinc** [comment]
Now we can worry about all the differences in https://github.com/eic/EICrecon/actions/runs/6712073374/job/18241252561?pr=1076#step:11:76. Are they all insignifcant or are there significant differences hidden between the insignificant ones?

**Comment by ShujieL** [comment]
> Now we can worry about all the differences in https://github.com/eic/EICrecon/actions/runs/6712073374/job/18241252561?pr=1076#step:11:76. Are they all insignifcant or are there significant differences hidden between the insignificant ones?

Those differences are mostly index changes. Real value change (diff>1%)is very rare. I'd say it's insignificant

**Comment by wdconinc** [comment]
The problem is that I don't understand why these changes have ANY impact at all. Where are the changed lines that you can point to that legitimately would lead to change indices?

**Comment by ShujieL** [comment]
> The problem is that I don't understand why these changes have ANY impact at all. Where are the changed lines that you can point to that legitimately would lead to change indices?

We removed TrackerSourcelinkerResult, so the input collections to CKFtracking is now two instead of three.  I suspect those collection ID changes are simply due to the output collection order change in JEventProcessorPodio.

**Comment by wdconinc** [comment]
There is a difference between collection IDs (*identifiers*) and the index changes, for example
```
CalorimeterTrackProjections.points_end
  NOT EQUAL
  ('Value of root[36][5] changed from 19 to 20.\n'
   'Value of root[36][6] changed from 23 to 24.\n'
   'Value of root[36][7] changed from 27 to 28.\n'
   'Value of root[36][8] changed from 31 to 32.\n'
etc
```
These index changes indicate that the number of track points in the track projections changed for event 36 (there are more of them for the 5th track projection). That is what we need to understand and explain.

**Comment by simonge** [comment]
Might the differences all stem from the change from the acts Vector (from the GlobaToLocal) having double members and the Measurement2D location being a float vector? Some subtle change in the location adds or removes a point from the projection.

**Comment by wdconinc** [comment]
That could be. Would have to be evaluated locally.

**Comment by wdconinc** [comment]
Compared with #1098, only the measurement indices are different (and collectionIDs). Thanks @simonge for the hint. Thanks @ShujieL for the patience ;-)
```console
$ bara --serve -M '.*\#.' old/rec_dis_18x275_minQ2\=1_craterlake.edm4eic.root new/rec_dis_18x275_minQ2\=1_craterlake.edm4eic.root 
CentralCKFSeededTrajectories/CentralCKFSeededTrajectories.measurementHits_begin
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, ..., 0, 0], ..., [0]] [[0, 5, 10, 12, 17, 20, 23, 26, 31, 36, 41, 42, 44, 46, 49], [0, ...], ..., [0]]
CentralCKFSeededTrajectories/CentralCKFSeededTrajectories.measurementHits_end
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, ..., 0, 0], ..., [0]] [[5, 10, 12, 17, 20, 23, 26, 31, 36, 41, 42, 44, 46, 49, 52], [...], ..., [5]]
CentralCKFSeededTrajectories/CentralCKFSeededTrajectories.outlierHits_begin
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, ..., 0, 0], ..., [0]] [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 10], [0, 1, ..., 19], ..., [0]]
CentralCKFSeededTrajectories/CentralCKFSeededTrajectories.outlierHits_end
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, ..., 0, 0], ..., [0]] [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 10, 10], [1, 5, ..., 19], ..., [0]]
CentralCKFTrajectories/CentralCKFTrajectories.measurementHits_begin
[[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, ..., 0, 0, 0, 0, 0], ..., [0, 0, 0, 0, 0]] [[0, 5, 8, 13, 14, 17], [0, 5, 9, 13, ..., 37, 41, 42], ..., [0, 5, 6, 10, 14]]
CentralCKFTrajectories/CentralCKFTrajectories.measurementHits_end
[[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, ..., 0, 0, 0, 0, 0], ..., [0, 0, 0, 0, 0]] [[5, 8, 13, 14, 17, 18], [5, 9, 13, ..., 41, 42, 43], ..., [5, 6, 10, 14, 15]]
CentralCKFTrajectories/CentralCKFTrajectories.outlierHits_begin
[[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, ..., 0, 0, 0, 0, 0], ..., [0, 0, 0, 0, 0]] [[0, 1, 2, 3, 4, 4], [0, 2, 4, 4, 4, ..., 12, 12, 12, 12], ..., [0, 0, 0, 0, 0]]
CentralCKFTrajectories/CentralCKFTrajectories.outlierHits_end
[[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, ..., 0, 0, 0, 0, 0], ..., [0, 0, 0, 0, 0]] [[1, 2, 3, 4, 4, 6], [2, 4, 4, 4, 7, ..., 12, 12, 12, 12], ..., [0, 0, 0, 0, 0]]
```

**Review by wdconinc — APPROVED** [review]

**Comment by ShujieL** [comment]
> That could be. Would have to be evaluated locally.

I pulled out the raw log and turn it into a data frame. There are 281k entries of value difference, among those:
* 106k entries are collectionID which is understood.
* 6k entries are index changes. Most of them are changes from 0 to some meaningful index for measurementHits. A few of them (<100 entries) associated with change in CalorimeterTrackProjection surfaces of two events (out of 100 events).
* for the remained 168k of value changes, 13k of them are change greater than 0.1%, among those, 12k are from DRICHGasIrtCherenkovParticleID_1.a (or _1.b). 

That is all I can come up with now. Will leave it to the reviewer to decide how to proceed.
